File: edition4/js/ch02/2-4.js
function Node(data, next) {
  this.data = data;
  this.next = next;
};

function listToNumber(head) {
  var num = 0;
  var place = 0;
  var curr = head;

  while (curr) {
    num += curr.data * (Math.pow(10, place));

    curr = curr.next
    place += 1;
  }

  return num;
}

function add(head1, head2) {
  return listToNumber(head1) + listToNumber(head2);
}

// tests
var assert = require('assert');

describe('add', function() {

  it('should return 0 when given to null pointers', function() {
    var actual = add(null, null);
    assert.equal(actual, 0);
  });

  it('should add two numbers when given a null point and one list', function() {
    var n2 = new Node('3', null);
    var n1 = new Node('2', n2);
    var head = new Node('1', n1);

    var actual = add(head, null);
    assert.equal(actual, 321);
  });

  it('should add two numbers when given two lists', function() {
    var n2 = new Node('3', null);
    var n1 = new Node('2', n2);
    var head1 = new Node('1', n1);

    var n3 = new Node('3', null);
    var n4 = new Node('2', n2);
    var head2 = new Node('1', n1);

    var actual = add(head1, head2);
    assert.equal(actual, 642);
  });

});

File: edition4/c/ch02/2-4.cpp
#include <assert.h>

#include <algorithm>
#include <cmath>
#include <iostream>
#include <string>

using namespace std;

class Node {
  public:
    int value;
    Node *next;
    Node(int value, Node *next);
};

Node::Node(int value, Node *next) {
  this->value = value;
  this->next = next;
}

class LinkedList {
  public:
    Node *head;
    LinkedList();
    void add(int value);
    void deleteFromNode(Node *n);
    int length();
    void print();
};

LinkedList::LinkedList() {
  this->head = NULL;
}

void LinkedList::add(int value) {
  if (!this->head) {
    this->head = new Node(value, NULL);
  } else {
    Node *temp = this->head;

    while (temp->next) {
      temp = temp->next;
    }

    temp->next = new Node(value, NULL);
  }
}

int LinkedList::length() {
  int length = 0;
  Node *curr = this->head;

  while (curr) {
    length++;
    curr = curr->next;
  }

  return length;
}

void LinkedList::print() {
  Node *curr = this->head;

  while (curr) {
    cout << curr->value << " -> ";
    curr = curr->next;
  }

  cout << "NULL" << endl;
}

int listToInt(LinkedList list) {
  int n = 0;
  int multiplier = 1;
  Node *curr = list.head;

  while (curr) {
    n += curr->value * multiplier;

    curr = curr->next;
    multiplier *= 10;
  }

  return n;
}

int sumLists(LinkedList n1, LinkedList n2) {
  return listToInt(n1) + listToInt(n2);
}

int main() {
  LinkedList n1;
  n1.add(1);
  n1.add(2);
  n1.add(3);
  LinkedList n2;
  n2.add(1);
  n2.add(2);
  n2.add(3);
  assert(sumLists(n1, n2) == 642);

  LinkedList n3;
  LinkedList n4;
  n4.add(2);
  n4.add(4);
  assert(sumLists(n3, n4) == 42);

  return 0;
}
